### PR TITLE
qa: Use node.datadir instead of tmpdir in test framework

### PR DESCRIPTION
**Pull request by maflcko**
Commit c53c9831eedaf3b311bb942945268830f9ba3abc introduced the utility function `get_datadir_path`, however not all places in the code use this util function. Using the util function everywhere makes it easier to review pull requests related to the datadir.

This commit replaces datadir path creation with the `datadir` member of `TestNode`, which itself uses `get_datadir_path`.

### CODE REVIEW

File: test/functional/keypool-topup.py [hunk 1]
```diff
@@ -10,63 +10,57 @@
...
-        self.tmpdir = self.options.tmpdir
+        datadir_regtest_1 = os.path.join(get_datadir_path(self.options.tmpdir, 1), 'regtest')
         self.nodes[0].generate(101)
 
         self.log.info("Make backup of wallet")
-
         self.stop_node(1)
-
-        shutil.copyfile(self.tmpdir + "/node1/regtest/wallets/wallet.dat", self.tmpdir + "/wallet.bak")
+        shutil.copyfile(datadir_regtest_1 + "/wallets/wallet.dat", datadir_regtest_1 + "/wallet.bak")
```
Comment (promag) [inline test/functional/keypool-topup.py]: Should also use `os.path.join`?

**Review by promag — COMMENTED** [review]
utACK fa72ada.

BTW, maybe this should be simpler:
```py
foo = self.node_path(1, 'regtest', 'wallets', 'wallet.dat')
```

**Comment by herojab** [comment]
os.path.join

File: test/functional/keypool-topup.py (hunk 1 above)
Comment (markaw67) [reply test/functional/keypool-topup.py]: Hello how are you!!

**Comment by promag** [comment]
utACK fa3dda1, will check for missing changes.

File: test/functional/test_framework/test_framework.py [hunk 2]
```diff
@@ -213,7 +217,7 @@ def add_nodes(self, num_nodes, extra_args=None, rpchost=None, timewait=None, bin
         assert_equal(len(extra_args), num_nodes)
         assert_equal(len(binary), num_nodes)
         for i in range(num_nodes):
-            self.nodes.append(TestNode(i, self.options.tmpdir, extra_args[i], rpchost, timewait=timewait, binary=binary[i], stderr=None, mocktime=self.mocktime, coverage_dir=self.options.coveragedir))
+            self.nodes.append(TestNode(i, self.node_path(i), extra_args[i], rpchost, timewait=timewait, binary=binary[i], stderr=None, mocktime=self.mocktime, coverage_dir=self.options.coveragedir))
```
Comment (promag) [inline test/functional/test_framework/test_framework.py]: Use `get_datadir_path(self.options.tmpdir, i)` instead.

File: test/functional/test_framework/test_framework.py [hunk 3]
```diff
@@ -203,6 +204,9 @@ def run_test(self):
 
     # Public helper methods. These can be accessed by the subclass test scripts.
 
+    def node_path(self, i, *paths):
+        return os.path.join(get_datadir_path(self.options.tmpdir, i), *paths)
```
Comment (promag) [inline test/functional/test_framework/test_framework.py]: Use `os.path.join(node[i].datadir, *paths)` instead.

**Review by promag — COMMENTED** [review]
With these suggestions, `node_path` honours runtime changes to `datadir`, not to mention `get_datadir_path` is called once only.

**Review by promag — COMMENTED** [review]
Nit, there are a couple of these (`node[i].datadir`) that could be changed:
```diff
-        assert os.path.isfile(os.path.join(self.nodes[0].datadir, "regtest", "debug.log"))
+        assert os.path.isfile(self.node_path(0, 'regtest', 'debug.log'))
```

**Comment by jnewbery** [comment]
concept ACK.

What are the benefits of using `self.node_path(i)` over `self.nodes[i].datadir`?

**Comment by promag** [comment]
Implicit path join.

**Comment by jnewbery** [comment]
ok, let me rephrase: what are the benefits of implementing this as a method on `BitcoinTestFramework` rather than on `TestNode`?

**Comment by promag** [comment]
@jnewbery I guess either are ok. You prefer `node[i].path()` or something like that? My suggestion was to avoid `os.path.join()` or `datadir + '/' foo` all over the tests.
